### ISSUE TITLE
ts_transformer_keys_1.keys is not a function

### ISSUE THREAD

**Issue by llanginger** [kimamula/ts-transformer-keys#4]
Hi there!

This package sounds like the ideal solution to a problem I have but I am struggling to get past the following error: "TypeError: ts_transformer_keys_1.keys is not a function" and I am not sure how to tell if this is an error on my part or if there is something out of my control that is not working.

I have tried compiling the following file with typescript 2.4.0 and running in the browser with no other packages:

```
import { keys } from "ts-transformer-keys";

interface Props {
    id: string;
    name: string;
    age: number;
}
const keysOfProps = keys<Props>();

console.log(keysOfProps); 
```

I have also tried using ts-node and executing the same file from the command line, though here in particular I am unsure to what extend ts-node is a bottleneck.

I hope I'm missing something super simple! 

Thanks in advance for any thoughts in any direction!

**Comment by llanginger** [comment]
I think I misunderstood something fundamental about how one is meant to use this! Will reopen if after more thorough testing I still can't get anywhere!

**Comment by hhshan** [comment]
@llanginger were you able to come up with a workaround. cause I too ran into the same issue trying to use this package.

**Comment by kimamula** [comment]
The error `TypeError: ts_transformer_keys_1.keys is not a function` means that you have failed to get the custom transformer working during compilation.

Could you show me how you are using the custom transformer to compile your TypeScript codes?

**Comment by hhshan** [comment]
Hi @kimamula I got around the issue.. was a problem with how i used the transformer. thanks for the library.. helps to solve a major issue i was facing with

**Comment by martnst** [comment]
I get the same `TypeError: ts_transformer_keys_1.keys is not a function` by simply adding it via `yarn add -D ts-transformer-keys` and then trying to use it as of the example. It seems there is some step missing in the docs?

**Comment by kimamula** [comment]
@martnst Sorry for the late response.
If you are using Windows, the issue is probably fixed by [this commit](https://github.com/kimamula/ts-transformer-keys/pull/10/commits/8c6145e1735278280ed5c3a2d56e0c27a31caf91).
Please try v0.3.1.

**Comment by macbric** [comment]
@kimamula  0.3.1 does not work on Windows.

**Comment by martnst** [comment]
FYI: I am on macOS

**Comment by kimamula** [comment]
@macbric @martnst I confirmed the following works both on macOS and Windows (with TypeScript@2.8.3).

* index.ts

```ts
import { keys } from 'ts-transformer-keys';

interface Props {
  id: string;
  name: string;
  age: number;
}
const keysOfProps = keys<Props>();

console.log(keysOfProps);
```

* compile.js

```js
const ts = require('typescript');
const keysTransformer = require('ts-transformer-keys/transformer').default;

const program = ts.createProgram(['./index.ts'], {
  strict: true,
  noEmitOnError: true,
  target: ts.ScriptTarget.ES5
});

const transformers = {
  before: [keysTransformer(program)],
  after: []
};
const { emitSkipped, diagnostics } = program.emit(undefined, undefined, undefined, false, transformers);

if (emitSkipped) {
  throw new Error(diagnostics.map(diagnostic => diagnostic.messageText).join('\n'));
}
```

* compile and run

```sh
$ node compile # creates index.js
$ node index   # outputs "[ 'id', 'name', 'age' ]"
```

Could you provide your code so that I can investigate the problem?

**Comment by ludogithub** [comment]
Hi,

For me the code : 

```
import { keys } from 'ts-transformer-keys';

interface Props {
  id: string;
  name: string;
  age: number;
}
const keysOfProps = keys<Props>();

console.log(keysOfProps);
```

Gets compiled as : 

```
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ts_transformer_keys_1 = require("ts-transformer-keys");
const keysOfProps = ts_transformer_keys_1.keys();
console.log(keysOfProps);
```

And serves the error : 

`TypeError: ts_transformer_keys_1.keys is not a function`

On windows, with typescript 5.6.0.

Best,

Ludovic

**Comment by kimamula** [comment]
Hi.
I have listed [several ways](https://github.com/kimamula/ts-transformer-keys#how-to-use-the-custom-transformer) of how to use custom transformers in README.
Which way are you using?

**Comment by tentreescantbeatme** [comment]
I see you added support for your plugin to awesome-typescript-loader. But unfortunately I switched to ts-loader some time ago because of performance problems: https://github.com/s-panferov/awesome-typescript-loader/issues/570

Any chance of added support in ts-loader?

**Comment by kimamula** [comment]
@tentreescantbeatme Sorry for the late reply.
At that time, I also looked into ts-loader implementation and gave up to add support to it as it looked it is not easy to provide `getCustomTransformers()` with `Program` instance.

There is [an issue](https://github.com/TypeStrong/ts-loader/issues/860) on the ts-loader repository, so someone may working on it.
You can use ttypescript for the meantime.

**Comment by kimamula** [comment]
@tentreescantbeatme [this issue](https://github.com/TypeStrong/ts-loader/issues/860) was fixed and `ts-loader@5.3.3` can provide `ts.Program` to custom transformers, which enables `ts-transformer-keys` to work with `ts-loader`.

**Comment by Thanos-Pappas** [comment]
Hello,
I'm getting the following error. `TypeError: Object(...) is not a function` on this line `console.log( keys<MyInterface>()[0]);`
I am using the following versions :
- Angular: 7.2.0
- typescript: 3.2.2
- webpack: 4.28.4
- awesome-typescript-loader: 5.2.0
- @angular-builders/custom-webpack: 7.2.0
- @angular-builders/dev-server: 7.2.1

A simplified version of my code:

my-inteface.ts
```
export interface MyInterface{
  keyA: string;
  keyB: string;
  keyC: string;
}
```

testTransformer.ts
```
import {keys} from 'ts-transformer-keys';
import {MyInterface} from '../../../interfaces/my-inteface';

export class TestTransformer{
  console.log( keys<MyInterface>()[0]);
}
```

I followed the webpack configuration as described [here](https://codeburst.io/customizing-angular-cli-6-build-an-alternative-to-ng-eject-a48304cd3b21)

My configurations follow:

angular.json
```
...
"architect": {
        "build": {
          "builder": "@angular-builders/custom-webpack:browser",
          "options": {
            "customWebpackConfig": {
              "path": "./extra-webpack.config.js"
            },
...

"serve": {
          "builder": "@angular-builders/dev-server:generic",
...
```

extra-webpack.config.js
```
const keysTransformer = require('ts-transformer-keys/transformer').default;
module.exports = {
  module: {
    rules: [
      {
        test: /\.ts$/,
        loader: 'awesome-typescript-loader',
        options: {
          getCustomTransformers: program => ({
  before: [
    keysTransformer(program)
            ]
          })
        }
      }
    ]}};
```
and I run the application with `ng serve`

**Any ideas what is going wrong here?**

**Comment by gabbifish** [comment]
I am also using the same default webpack config as @Thanos-Pappas and am running into:
```
TypeError: ts_transformer_keys_1.keys is not a function
```
is there something missing from the webpack demo code?

**Comment by zucker-rdrg** [comment]
I have same problem with `jasmine-ts` and `ts-transformer-keys`

**Comment by kimamula** [comment]
@Thanos-Pappas I tested `@angular-builders/custom-webpack` with `ts-transformer-keys` and confirmed `TypeError: Object(...) is not a function` error.

I found several issues (such as [this](https://github.com/meltedspark/angular-builders/issues/186) and [this](https://github.com/meltedspark/angular-builders/issues/179)) suggesting that you have to use `mergeStrategies: { "module": "prepend" }` to apply additional TypeScript transformations with `@angular-builders/custom-webpack`.
However, using `mergeStrategies: { "module": "prepend" }` results in compilation errors such as the following:

```
ERROR in [at-loader] ./src/app/app.component.ts:8:5 
    TS2539: Cannot assign to 'AppComponent' because it is not a variable.
```

This even happens when I use `awesome-typescript-loader` (or `ts-loader`) without custom transformers, e.g., like the following:

```js
module.exports = {
  module: {
    rules: [
      {
        test: /\.ts$/,
        loader: 'awesome-typescript-loader' // or 'ts-loader'
      }
    ]
  }
};
```

It is likely that

1. TypeScript transformations executed before angular transformers' transformation are ignored (`mergeStrategies: { "module": "append" }`, which is default)
2. TypeScript transformations with `awesome-typescript-loader` or `ts-loader` are not possible after angular transformers' transformation because the TS code is no longer compilable (`mergeStrategies: { "module": "prepend" }`)

Note that [webpack loaders are applied from the bottom to the top](https://stackoverflow.com/questions/32234329/what-is-the-loader-order-for-webpack).

As of now, I cannot think of any good solution for this issue.

@gabbifish I haven't encountered `TypeError: ts_transformer_keys_1.keys is not a function` with `@angular-builders/custom-webpack` but this may be also relate to you.

**Comment by simllll** [comment]
For me the issue was that I used the fork-ts-checker-webpack-plugin and therefore set  transpileOnly: true. If transpileOnly is true, this plugin does not work. Makes kinda sense, took me still quite a while to figure it out ;).

TLDR: ensure transpileOnly is not set to true!

Regards
Simon

**Comment by ipetrovic11** [comment]
Hi @kimamula ,

Have you found the way to make this work with Angular or @ngtools/webpack?

The best I was able to do is to patch @ngtools/webpack so it uses ttypescript, but not sure that is sustainable model.

**Comment by pedrogscruz** [comment]
> I had the same problem i had to follow it using **ttypescript**.
> 
> 1. `npm install --save-dev typescript ttypescript`
> 2. in **package.json** on `"start":` use `ttsc` instead of `tsc`
> 
> ```
> {
>   "name": "api",
>   "version": "0.0.0",
>   "scripts": {
>     "start": "ttsc && node ./dist/server",
>     "test": "echo \"Error: no test specified\" && exit 1"
>   },
>   "dependencies": {
>     ...
>   },
>   "devDependencies": {
>     ...
>   }
> ```
> 
> 1. in **tsconfig.json** include plugin -> transform
> 
> ```
> {
>   "compilerOptions": {
>     ...
>     "plugins": [
>       { 
>         "transform": "ts-transformer-keys/transformer"
>       }
>     ]
>   },
>   ...
> }
> ```
> 
> I stopped getting the error after that

I did exactly the same thing here and it still does not work.
My React App get the error TypeError: Object(...) is not a function.

**Comment by RyoIkarashi** [comment]
I'm in the exact same scenario above.
Any solution or workaround on this issue?

**Comment by AmitSharmamad** [comment]
There is no such function as keys in the index.js file, kindly check that

**Comment by odykyi** [comment]
I have the same error TypeError: ts_transformer_keys_1.keys is not a function

I do not have webpack 

I have only  tsconfig.json and package.json 

How I can fix this error?

**Comment by kimamula** [comment]
How are you transforming `keys()`?
As listed in [README](https://github.com/kimamula/ts-transformer-keys#how-to-use-the-custom-transformer), there are typically 4 ways to transform `keys()`.

**Comment by dungnguyen10989** [comment]
> How are you transforming `keys()`?
> As listed in [README](https://github.com/kimamula/ts-transformer-keys#how-to-use-the-custom-transformer), there are typically 4 ways to transform `keys()`.

I'm using TypeScript with React-Native, and error `ts_transformer_keys_1.keys` still appears, this is my devDendencies: 
```
		"babel-jest": "^24.9.0",
		"babel-plugin-module-resolver": "^4.0.0",
		"eslint": "^6.5.1",
		"eslint-import-resolver-babel-module": "^5.1.2",
		"eslint-plugin-import": "^2.20.2",
		"eslint-plugin-module-resolver": "^0.16.0",
		"jest": "^24.9.0",
		"metro-react-native-babel-preset": "^0.58.0",
		"react-native-schemes-manager": "^2.0.0",
		"react-test-renderer": "16.11.0",
		"reactotron-react-native": "^5.0.0",
		"reactotron-redux": "^3.1.3",
		"reactotron-redux-saga": "^4.2.3",
		"ts-transformer-keys": "0.4.1",
		"tslint": "^6.1.2",
		"ttypescript": "^1.5.10",
		"typescript": "^3.8.3"
```
and `tsconfig.json`: 
```
{
	"compilerOptions": {
		"allowJs": true,
		"allowSyntheticDefaultImports": true,
		"esModuleInterop": true,
		"isolatedModules": false,
		"jsx": "react",
		"lib": ["es6", "dom", "es2017"],
		"moduleResolution": "node",
		"noEmit": true,
		"strict": true,
		"target": "esnext",
		"baseUrl": ".",
		"paths": {
			"*": ["src"]
		},
		"plugins": [{ "transform": "ts-transformer-keys/transformer" }]
	},
	"exclude": [
		"node_modules",
		"babel.config.json",
		"metro.config.json",
		"jest.config.json"
	]
}

```
How can i fix it?

**Comment by ArashMotamedi** [comment]
In case this helps anybody else: I was planning to use `ttsc` but had mistakenly pasted the `compilerOptions` object into my `package.json`. The correct place is inside `tsconfig.json` file.

**Comment by dungnguyen10989** [comment]
> In case this helps anybody else: I was planning to use `ttsc` but had mistakenly pasted the `compilerOptions` object into my `package.json`. The correct place is inside `tsconfig.json` file.

Can you show your `tsconfig.json` and `package.json` ? I've config like document, but this uncomfortable error still appears.

**Comment by MaddoxDevelopment** [comment]
I was able to get this working with Create React App by using https://github.com/timarney/react-app-rewired and https://www.npmjs.com/package/ttypescript

config-overrides.js
```
module.exports = function override(config, env) {
    config.module.rules.push(  {
        test: /\.(ts|tsx)$/,
        loader: require.resolve('awesome-typescript-loader'),
        options: {
            compiler: 'ttypescript'
        }
    })
    return config;
}
```

tsconfig.json
```
{
  "compilerOptions": {
    "target": "es5",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": false,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "preserve",
    "plugins": [
      {
        "transform": "ts-transformer-keys/transformer"
      }
    ]
  },
  "include": [
    "src"
  ],
  "exclude": [
    "node_modules",
    "src/jestTrxProcessor.js",
    "src/test-*.js",
    "src/setupTests.js"
  ]
}
```

**Comment by kimamula** [comment]
@dungnguyen10989 If you are using Babel, it is not possible to get the transformer work. See #32.

**Comment by vmirage** [comment]
I have this issue as well with pnpm. It seems that the issue here is the filepath mismatch due to symbolic link created by pnpm.

It can be fixed by changing

https://github.com/kimamula/ts-transformer-keys/blob/8a92a56719e97e1e849de5670148f09ff63aaf3f/transformer.ts#L61

to

```javascript
var fs = require('fs');

...

 && (fs.realpathSync(path.join(declaration.getSourceFile().fileName)) === indexTs)
```

Not sure if this is the right way to fix it. But it works.

**Comment by kimamula** [comment]
@vmirage Thanks for reporting! I will look for if there is a better way to resolve this problem.

**Comment by kimamula** [comment]
@vmirage I found that using `require.resolve()` instead of `path.join()` works and published the fixed version (v0.4.2). Please give it a try. Thanks again for your report!

**Comment by vmirage** [comment]
@kimamula yup that works

**Comment by luisdemarchi** [comment]
@kimamula  I arrived late, but unfortunately in the same bug. Using version 0.4.2 

TypeError: ts_transformer_keys_1.keys is not a function

**Comment by kimamula** [comment]
@luisdemarchi please provide a reproduction. Thanks!

**Comment by benjamingwynn** [comment]
After testing this all day I think neither ts-transformer-keys or [ts-interface-keys-transformer](https://github.com/nullcc/ts-interface-keys-transformer) work with Webpack [ts-loader](https://github.com/TypeStrong/ts-loader) when `onlyCompiledBundledFiles` is enabled. It appears as though the typechecker in the transformer code cannot determine the declaration from the symbol because the file is technically not imported as it's not one of the aforementioned "compiled bundle files" (`typeChecker.getResolvedSignature(node).signature.declaration` in `isKeysCallExpression` always reads undefined.) 

This _kind of_ makes sense, and I've disabled `onlyCompiledBundledFiles` for now, although that isn't ideal for my production build at all. Maybe somehow manually including the declaration file in the bundled files would fix it (I've tried using `include` on the webpack ts-loader module rule but to no avail, I'm still learning how to intelligently configure Webpack ). There are some mentions that .d.ts files will be automatically picked up, but the documentation is somewhat scarce (https://github.com/TypeStrong/ts-loader#onlycompilebundledfiles)

the following seem related: https://github.com/TypeStrong/ts-loader/issues/1002, https://stackoverflow.com/a/52034546/3754229

**Comment by kimamula** [comment]
@benjamingwynn The [doc](https://github.com/TypeStrong/ts-loader/blob/master/README.md#onlycompilebundledfiles) of ts-loader says:

> The onlyCompileBundledFiles option modifies this behavior, loading only those files that are actually bundled by webpack, as well as **any .d.ts files included by the tsconfig.json settings**.

And I confirmed that adding `"files": ["node_modules/ts-transformer-keys/index.d.ts"],` in `tsconfig.json` solves the problem. Please give it a try in your build.

**Comment by robbiemu** [comment]
it is working for me with ttypecript, but in ts-jest I am getting this error.

**Comment by stevensnoeijen** [comment]
Got the same problem when running it in vscode  with jest-runner, `jest --clearCache` fixed this problem :)

**Comment by alexlbr** [comment]
I had to update the [factory function](https://github.com/kulshekhar/ts-jest/blob/master/src/types.ts#L234) to make it work using `"ts-jest": "^27.0.4"` 
```js
// notice cs.program instead of cs.tsCompiler.program
const factory = cs => ctx => keysTransformer(cs.program)(ctx); // 👈 no `tsCompiler`
```

**Comment by ahmadalibaloch** [comment]
Using nodejs with ts-node as `ts-node ./src/server.ts` and getting `TypeError: (0 , ts_transformer_keys_1.keys) is not a function`

**Comment by arnav7633** [comment]
Was this fixed?
I am getting the same issue

**Comment by keithort** [comment]
For anyone still struggling with this, another fun wrinkle I ran into was that the uses of `keys` that lived in `__tests__` would throw this error. if I moved my tests into the root, then the function was found and executed.

**Comment by Ghostdog6** [comment]
We had been using this succesfully, but were having a similar problem when attempting to update our libraries (especially node).  I'd updated my devDependencies to use "typescript": "^4.2.3".  The server would build, but when trying to use the keys function we'd get an error saying "ts_transformer_keys_1 is not defined".

A coworker noticed that the devDependency on this package was ^3.9.5, so I did the same locally and it works again.

**Comment by behemoth-thainv** [comment]
> I had the same problem i had to follow it using **ttypescript**.
> 
> 1. `npm install --save-dev typescript ttypescript`
> 2. in **package.json** on `"start":` use `ttsc` instead of `tsc`
> 
> ```
> {
>   "name": "api",
>   "version": "0.0.0",
>   "scripts": {
>     "start": "ttsc && node ./dist/server",
>     "test": "echo \"Error: no test specified\" && exit 1"
>   },
>   "dependencies": {
>     ...
>   },
>   "devDependencies": {
>     ...
>   }
> ```
> 
> 3. in **tsconfig.json** include plugin -> transform
> 
> ```
> {
>   "compilerOptions": {
>     ...
>     "plugins": [
>       { 
>         "transform": "ts-transformer-keys/transformer"
>       }
>     ]
>   },
>   ...
> }
> ```
> 
> I stopped getting the error after that 😅

It's work for me!

**Comment by nmgix** [comment]
nest.js + basic Jest testing
ts-jest + docs helped

**jest.json**
```
{
  "moduleFileExtensions": ["js", "json", "ts"],
  "rootDir": "../",
  "testEnvironment": "node",
  "testRegex": ".spec.ts$",
  "transform": {
    "^.+\\.(t|j)s$": "ts-jest"
  },
  "moduleNameMapper": {
    "^src/(.*)$": "<rootDir>/../src/$1"
  },
  "preset": "ts-jest",
  "globals": {
    "ts-jest": {
      "astTransformers": { "before": ["testing/helpers/ts-jest-keys-transformer.js"] }
    }
  }
}
``` 
and added [ts-jest-keys-transformer.js](https://github.com/kimamula/ts-transformer-keys/blob/master/examples/ts-jest/ts-jest-keys-transformer.js) in same folder

issue is that it warns `ts-jest[ts-jest-transformer] (WARN) Define 'ts-jest' config under 'globals' is deprecated.`
